ext/tls/tls_key.rs

```diff
@@ -0,0 +1,320 @@
+          Ok(Ok(res)) => Ok(res),
+        })
+        .boxed()
+    })
+  }
+}
+
+impl Debug for TlsKeyResolver {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    f.debug_struct("TlsKeyResolver").finish()
+  }
+}
+
+pub fn new_resolver() -> (TlsKeyResolver, TlsKeyLookup) {
+  let (resolution_tx, resolution_rx) = mpsc::unbounded_channel();
+  (
+    TlsKeyResolver {
+      inner: Rc::new(TlsKeyResolverInner {
+        resolution_tx,
+        cache: Default::default(),
+      }),
+    },
+    TlsKeyLookup {
+      resolution_rx: RefCell::new(resolution_rx),
+      pending: Default::default(),
+    },
+  )
+}
+
+impl TlsKeyResolver {
+  /// Resolve the certificate and key for a given host. This immediately spawns a task in the
+  /// background and is therefore cancellation-safe.
+  pub fn resolve(
+    &self,
+    sni: String,
+  ) -> impl Future<Output = Result<TlsKey, AnyError>> {
+    let mut cache = self.inner.cache.borrow_mut();
+    let mut recv = match cache.get(&sni) {
+      None => {
+        let (tx, rx) = broadcast::channel(1);
+        cache.insert(sni.clone(), TlsKeyState::Resolving(rx.resubscribe()));
+        _ = self.inner.resolution_tx.send((sni.clone(), tx));
+        rx
+      }
+      Some(TlsKeyState::Resolving(recv)) => recv.resubscribe(),
+      Some(TlsKeyState::Resolved(res)) => {
+        return Either::Left(ready(res.clone().map_err(|_| anyhow!("Failed"))));
+      }
+    };
+    drop(cache);
+
+    // Make this cancellation safe
+    let inner = self.inner.clone();
+    let handle = spawn(async move {
+      let res = recv.recv().await?;
+      let mut cache = inner.cache.borrow_mut();
+      match cache.get(&sni) {
+        None | Some(TlsKeyState::Resolving(..)) => {
+          cache.insert(sni, TlsKeyState::Resolved(res.clone()));
+        }
+        Some(TlsKeyState::Resolved(..)) => {
+          // Someone beat us to it
+        }
+      }
+      res.map_err(|_| anyhow!("Failed"))
+    });
+    Either::Right(async move { handle.await? })
+  }
+}
+
+pub struct TlsKeyLookup {
+  #[allow(clippy::type_complexity)]
+  resolution_rx: RefCell<
+    mpsc::UnboundedReceiver<(
+      String,
+      broadcast::Sender<Result<TlsKey, ErrorType>>,
+    )>,
+  >,
+  pending:
+    RefCell<HashMap<String, broadcast::Sender<Result<TlsKey, ErrorType>>>>,
+}
+
+impl TlsKeyLookup {
+  /// Only one poll call may be active at any time. This method holds a `RefCell` lock.
+  pub async fn poll(&self) -> Option<String> {
+    if let Some((sni, sender)) =
+      poll_fn(|cx| self.resolution_rx.borrow_mut().poll_recv(cx)).await
+    {
+      self.pending.borrow_mut().insert(sni.clone(), sender);
+      Some(sni)
+    } else {
+      None
+    }
+  }
+
+  /// Resolve a previously polled item.
+  pub fn resolve(&self, sni: String, res: Result<TlsKey, AnyError>) {
+    _ = self
+      .pending
+      .borrow_mut()
+      .remove(&sni)
+      .unwrap()
```

Comment:
Should we gracefully error here?

Reply:
This would definitely be a serious bug if this was somehow called twice. I think it's probably fine if we panic.